File: js-template/leetcode/editor/cn/symmetric-tree.js
/*
 * @lc app=leetcode.cn id=101 lang=javascript
 * @lcpr version=30404
 *
 * [101] 对称二叉树
 */


import {ListNode} from "../common/listNode.js";
import {TreeNode} from "../common/treeNode.js";

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function(root) {
    return isMirror(root.left, root.right);
};

/**
 * @param {TreeNode} left 
 * @param {TreeNode} right 
 */
var isMirror = function(left, right) {
    if (left != null && right == null)
        return false;
    else if (left == null && right != null)
        return false;
    else if (left == null && right == null)
        return true;
    else if (left.val != right.val)
        return false;
    else
        return isMirror(left.left, right.right) && isMirror(left.right, right.left);
}


// @lc code=end

// your test code here

isSymmetric(TreeNode.createRoot([1,2,2,3,4,4,3]))


/*
// @lcpr case=start
// [1,2,2,3,4,4,3]\n
// @lcpr case=end

// @lcpr case=start
// [1,2,2,null,3,null,3]\n
// @lcpr case=end

 */

File: cpp-template/leetcode/editor/cn/symmetric-tree.cpp
/*
 * @lc app=leetcode.cn id=101 lang=cpp
 * @lcpr version=30404
 *
 * [101] 对称二叉树
 */

#include <iostream>
#include <vector>
#include <string>
#include <list>
#include "../common/ListNode.cpp"
#include "../common/TreeNode.cpp"

using namespace std;

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isMirror(TreeNode *left, TreeNode *right) {
        if ((!left && right) || (left && !right))
            return false;
        if (!left && !right)
            return true;
        return left->val == right->val && isMirror(left->left, right->right) && isMirror(left->right, right->left);
    }

    bool isSymmetric(TreeNode* root) {
        return isMirror(root->left, root->right);
    }
};
// @lc code=end

int main() {
    Solution solution;
    // your test code here
    TreeNode *root = TreeNode::createRoot({1,2,2,2,0,2});
    solution.isSymmetric(root);
}



/*
// @lcpr case=start
// [1,2,2,3,4,4,3]\n
// @lcpr case=end

// @lcpr case=start
// [1,2,2,null,3,null,3]\n
// @lcpr case=end

// @lcpr case=start
// [1,2,2,2,null,2]\n
// @lcpr case=end

 */

